runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -212,71 +358,119 @@ impl Module {
         F: FnOnce(
             &mut StackExecutor<'static, MemoryStackState<'_, 'static, evm_backend::Backend<'_, C>>>,
         ) -> (evm::ExitReason, V),
-        C: Context,
+        C: TxContext,
     {
-        // Make sure that gas_price >= min_gas_price.
         let params = Self::params(ctx.runtime_state());
-        let gas_price: primitive_types::U256 = gas_price_in.into();
+        let den = params.token_denomination;
         let min_gas_price: primitive_types::U256 = params.min_gas_price.into();
+
+        let gas_price: primitive_types::U256 = gas_price_in.into();
+
+        // Make sure that gas_price >= min_gas_price.
         if gas_price < min_gas_price {
             return Err(Error::GasPriceTooLow);
         }
 
+        // Make sure we have enough core gas left.
+        if core::Module::remaining_tx_gas(ctx) < gas_limit {
```

Comment:
Shouldn't it instead compute `gas_limit` from `remaining_tx_gas`? Or why is this separate `gas_limit` parameter needed?